### PR TITLE
Refactor ak #3934 turn lanes

### LINKED ISSUE #3934: Dual-use lane icons are highlighted incorrectly
In the lane view that appears below the turn banner, the icon for a dual use lane – a lane that can be used for multiple maneuvers – is highlighted incorrectly.

## Problem

If the user is supposed to turn left at an intersection and one of the lanes can be used for going straight or turning right, the lane icon is highlighted as if the user is supposed to turn right from that lane. It should not be highlighted at all.

## Diagnosis

`TurnLaneViewData.buildDrawData()` takes the lane indications for a given lane along with the overall maneuver that the lane may or may not be used for. In this case, `laneIndications` would be `STRAIGHT` and `RIGHT`, while `maneuverModifier` would be `LEFT`. However, the logic in `setDrawMethodWithModifier()` assumes the `maneuverModifier` has already been flipped just like the icon will be flipped, which is not the case. So `drawMethod` will be `DRAW_LANE_RIGHT_ONLY`, which looks like a straight-right icon. (`shouldFlip` will be `false` because the lane lacks a `LEFT` indication.)

https://github.com/mapbox/mapbox-navigation-android/blob/c5dc6c4238f5fd1e1bafd8e219dede40ab04c5d5/libnavigation-ui/src/main/java/com/mapbox/navigation/ui/internal/instruction/turnlane/TurnLaneViewData.java#L66-L72 https://github.com/mapbox/mapbox-navigation-android/blob/c5dc6c4238f5fd1e1bafd8e219dede40ab04c5d5/libnavigation-ui/src/main/java/com/mapbox/navigation/ui/internal/instruction/turnlane/TurnLaneViewData.java#L75-L83

By sheer coincidence, this is the same symptom seen in mapbox/mapbox-navigation-ios#2796. However, on iOS, the drawing logic was working correctly in this case, but the maneuver modifier was no longer being passed in as context, leaving a fallback pointing to the right.

## Next steps

`setDrawMethodWithModifier()` needs to consider whether the `maneuverModifier` matches one of the `laneIndications`. Conceptually, we should only highlight a lane use going in the same direction as the overall maneuver.

/cc @mapbox/navigation-android @avi-c

### PR DESCRIPTION
**Pull request by abhishek1508**
### Description

Fixes #3934 

### Changelog

```
<changelog>Refactored turn lane processing logic and added more lane combinations.</changelog>
```

### CODE REVIEW

File: libnavui-maneuver/src/main/java/com/mapbox/navigation/ui/maneuver/LaneIconProcessor.kt [hunk 1]
```diff
@@ -1,237 +1,310 @@
...
-                activeDirection?.let {
-                    if (it.contains(LEFT, ignoreCase = true)) {
-                        laneIcon[LaneTurns.LANE_LEFT_RIGHT_LEFT_ONLY]
-                    } else {
-                        laneIcon[LaneTurns.LANE_LEFT_RIGHT_RIGHT_ONLY]
-                    }
-                }
+            else -> {
+                // Draw "lane_straight"
+                LaneIcon(availableLaneNames["$LANE $STRAIGHT"]!!, shouldFlip)
```
Comment (LukasPaczos) [inline libnavui-maneuver/src/main/java/com/mapbox/navigation/ui/maneuver/LaneIconProcessor.kt:148]: As discussed, I still think that it's better to fall back to one of the actually possible turns for this lane (or a subset) than just hardcode "straight".

File: libnavui-maneuver/src/main/java/com/mapbox/navigation/ui/maneuver/LaneIconProcessor.kt [hunk 2]
```diff
@@ -1,237 +1,310 @@
...
-                        else -> {
-                            laneIcon[LaneTurns.LANE_SLIGHT_LEFT_STRAIGHT_SLIGHT_RIGHT_STRAIGHT_ONLY]
-                        }
-                    }
-                }
+            laneIndication.activeDirection != null -> {
+                // Draw "lane_laneIndication.activeDirection_using_laneIndication.activeDirection"
+                val active = laneIndication.activeDirection.replace(RIGHT, TURN).replace(LEFT, TURN)
+                val name = "$LANE $active $USING $active"
+                LaneIcon(availableLaneNames[name]!!, shouldFlip)
```
Comment (LukasPaczos) [inline libnavui-maneuver/src/main/java/com/mapbox/navigation/ui/maneuver/LaneIconProcessor.kt:144]: If there was a new lane indication added to the Directions API this code would start crashing for all clients. This is unlikely to happen in this particular scenario (it would be breaking change for the API), but let's take advantage of Kotlin and write a little bit more defensive code with fallbacks here.

File: libnavui-maneuver/src/main/java/com/mapbox/navigation/ui/maneuver/ManeuverProcessor.kt [hunk 3]
```diff
@@ -390,19 +405,28 @@ internal object ManeuverProcessor {
                     list.forEach {
                         val directions = it.directions()
                         val active = it.active()
+                        val activeDirection: String? = profile?.let { value ->
+                            // In case of PROFILE_CYCLING [BannerComponent#activeDirection] is
+                            // unavailable and hence we need to fallback to [BannerText#modifier]
+                            if (active!! && value == DirectionsCriteria.PROFILE_CYCLING) {
+                                bannerInstruction.primary().modifier()
```
Comment (LukasPaczos) [inline libnavui-maneuver/src/main/java/com/mapbox/navigation/ui/maneuver/ManeuverProcessor.kt:412]: As discussed, let's make this fallback dynamic depending on active direction availability and not hardcoding `PROFILE_CYCLING`.

File: libnavui-maneuver/src/main/java/com/mapbox/navigation/ui/maneuver/ManeuverProcessor.kt [hunk 4]
```diff
@@ -390,19 +405,28 @@ internal object ManeuverProcessor {
                     list.forEach {
                         val directions = it.directions()
                         val active = it.active()
+                        val activeDirection: String? = profile?.let { value ->
+                            // In case of PROFILE_CYCLING [BannerComponent#activeDirection] is
+                            // unavailable and hence we need to fallback to [BannerText#modifier]
+                            if (active!! && value == DirectionsCriteria.PROFILE_CYCLING) {
```
Comment (LukasPaczos) [inline libnavui-maneuver/src/main/java/com/mapbox/navigation/ui/maneuver/ManeuverProcessor.kt:411]: The assumption that `active` is present is only true for `Lane` component types which we're not checking above as far as I can tell.

File: libnavui-maneuver/src/main/java/com/mapbox/navigation/ui/maneuver/ManeuverProcessor.kt (hunk 3 above)
Comment (abhishek1508) [reply libnavui-maneuver/src/main/java/com/mapbox/navigation/ui/maneuver/ManeuverProcessor.kt:412]: The property `activeDirection` is a `Nullable` property irrespective of profile. This behavior to hook onto the `activeDirection` if available for any profile otherwise fallback to `primary.modifier` has a limitation in a sense that I cannot define `fun isActiveDirectionPresent()` as the only way to check if it exists is to check if it is null or not. Since it is a `Nullable` property this check is incorrect. For ex: look at below responses

##### Driving profile
```
"sub": {
    "components": [
        {
            "active_direction": "slight left",
            "active": true,
            "directions": [
                "slight left"
            ],
            "type": "lane",
            "text": ""
        },
        {
            "active": false,
            "directions": [
                "right"
            ],
            "type": "lane",
            "text": ""
        }
    ]
```

##### Cycling profile
```
"sub": {
    "components": [
        {
            "active": true,
            "directions": [
                "slight left"
            ],
            "type": "lane",
            "text": ""
        },
        {
            "active": false,
            "directions": [
                "right"
            ],
            "type": "lane",
            "text": ""
        }
    ]
```
Do you have any other ideas?
/cc @LukasPaczos

File: libnavui-maneuver/src/main/java/com/mapbox/navigation/ui/maneuver/ManeuverProcessor.kt (hunk 4 above)
Comment (abhishek1508) [reply libnavui-maneuver/src/main/java/com/mapbox/navigation/ui/maneuver/ManeuverProcessor.kt:411]: I didn't understand what you meant here?

File: libnavui-maneuver/src/main/java/com/mapbox/navigation/ui/maneuver/LaneIconProcessor.kt (hunk 1 above)
Comment (abhishek1508) [reply libnavui-maneuver/src/main/java/com/mapbox/navigation/ui/maneuver/LaneIconProcessor.kt:148]: This is very subjective and customers might react to it differently. Look at this comment here > https://github.com/mapbox/navigation-ui-resources/pull/26
```
I decided not to add canvases for the following lane configurations, which were exceedingly rare:

opposite slight turn or straight or sharp turn and opposite sharp turn or straight or slight turn (93 occurrences)
opposite sharp turn or slight turn and opposite slight turn or sharp turn (30 occurrences)
straight or turn or sharp turn (18 occurrences)
uturn or opposite slight turn (15 occurrences)
opposite turn or straight or sharp turn (12 occurrences)
straight or slight turn or sharp turn (8 occurrences)
straight or slight turn or uturn (3 occurrences)
opposite sharp turn or straight or sharp turn (3 occurrences)
straight or sharp turn or uturn (no occurrences)
turn or sharp turn or uturn (no occurrences)
```

As you can see the only combinations for which you would see a `lane_straight` is for the combination above which are very extremely rare. Some customers might like that we just say we don't know(by drawing straight) or some might like we are showing a subset. I think it's difficult to choose the behavior here and since these combinations are so rare, I feel it's ok to say we don't know. Also, let's take a design opinion here

/cc @d-prukop

Comment (LukasPaczos) [reply libnavui-maneuver/src/main/java/com/mapbox/navigation/ui/maneuver/LaneIconProcessor.kt:148]: I personally don't feel like showing straight indicates that we don't know. It might look like a much bigger mistake than at least drawing a single lane correctly.

If we'd like to be clear that we don't know, why not introduce a dedicated icon for it?

File: libnavui-maneuver/src/main/java/com/mapbox/navigation/ui/maneuver/LaneIconProcessor.kt (hunk 2 above)
Comment (LukasPaczos) [reply libnavui-maneuver/src/main/java/com/mapbox/navigation/ui/maneuver/LaneIconProcessor.kt:144]: > The value of activeDirection can never change and there is no possiblity that it can hold any other values than the one stated above.

That's right, this is our trust in the stability of the Directions API. That's what I'm saying wrt defensive programming - trust in documentation and stability of the API is one thing, but since we don't have a proper contract we might as well be on the safe side and verify the availability of the icon anyway since the language does give us this capability at a low cost.

File: libnavui-maneuver/src/main/java/com/mapbox/navigation/ui/maneuver/ManeuverProcessor.kt (hunk 3 above)
Comment (LukasPaczos) [reply libnavui-maneuver/src/main/java/com/mapbox/navigation/ui/maneuver/ManeuverProcessor.kt:412]: Why would this if-check not work?
```
if (active && active_direction == null) {
  bannerInstruction.primary().modifier()
}
```

File: libnavui-maneuver/src/main/java/com/mapbox/navigation/ui/maneuver/ManeuverProcessor.kt (hunk 4 above)
Comment (LukasPaczos) [reply libnavui-maneuver/src/main/java/com/mapbox/navigation/ui/maneuver/ManeuverProcessor.kt:411]: You're force-casting `active` from a nullable type to a non-nullable type here.

`active` can only be non-null if the component type is `Lane`, which you're not checking above. If some other component type ended up in this code path, we could crash. And again, with defensive programming in mind, why not check the nullability before using the type?

File: libnavui-maneuver/src/main/java/com/mapbox/navigation/ui/maneuver/LaneIconProcessor.kt (hunk 1 above)
Comment (abhishek1508) [reply libnavui-maneuver/src/main/java/com/mapbox/navigation/ui/maneuver/LaneIconProcessor.kt:148]: We have thought of `?` but we would need design input.

File: libnavui-maneuver/src/main/java/com/mapbox/navigation/ui/maneuver/ManeuverProcessor.kt (hunk 4 above)
Comment (abhishek1508) [reply libnavui-maneuver/src/main/java/com/mapbox/navigation/ui/maneuver/ManeuverProcessor.kt:411]: I am already checking if the type is `lane` using the if condition below
```
private fun getLaneGuidance(
        profile: String?,
        drivingSide: String,
        bannerInstruction: BannerInstructions
    ): Lane? {
        val subBannerText = bannerInstruction.sub()
        return subBannerText?.let { subBanner ->
            if (subBanner.type() == null && subBanner.text().isEmpty()) {
```
This is how the response was earlier. There was no `lane` type. Rather if the type was `null` we could say that it is a lane. I raised this as a bug a few months back and looks like directions fixed it. So now I need to add `StepManeuverType.Lane` to MapboxJava and then use it here. 

As far as `active` is concerned; this can never be null. I am not sure why is it annotated as `Nullable` in `Mapbox-Java`. Same goes for `LegStep.drivingSide`. Every field in MapboxJava is annoatated as nullable but if you look at directions documentation there are fields that can never be null. Also, if I check if (active == null), there is no defined behavior for it. This is a problem in general in our codebase I think, that we are making a lot of assumptions based on nullability from directions response and deciding the behavior, where is some cases we should just throw exceptions. 

So, I don't think there is any need for checking the nullability of `active` and I think I would just go forward with it.

File: libnavui-maneuver/src/main/java/com/mapbox/navigation/ui/maneuver/LaneIconProcessor.kt (hunk 2 above)
Comment (abhishek1508) [reply libnavui-maneuver/src/main/java/com/mapbox/navigation/ui/maneuver/LaneIconProcessor.kt:144]: What do we do if the name does not exist? In other words, if `active_direction` has a different value than the one stated above? 
```
laneIndication.activeDirection != null -> {
  // Draw "lane laneIndication.activeDirection using laneIndication.activeDirection"
  val active = laneIndication.activeDirection.replace(RIGHT, TURN).replace(LEFT, TURN)
  val name = "$LANE $active $USING $active"
  if (availableLaneNames.contains(name)) {
    LaneIcon(availableLaneNames[name]!!, shouldFlip)
  } else {
  // What do we return here?
  }
}
```

I honestly feel this would be a serious problem on the directions side if they assign a different value to the `active_direction` other than the ones stated above. Because, there cannot be any other possible lane you can take. However, both the platforms also need to align here. /cc @1ec5 for his inputs and how we want to deal with it.

File: libnavui-maneuver/src/main/java/com/mapbox/navigation/ui/maneuver/LaneIconProcessor.kt (hunk 1 above)
Comment (1ec5) [reply libnavui-maneuver/src/main/java/com/mapbox/navigation/ui/maneuver/LaneIconProcessor.kt:176]: First of all, keep in mind that this fallback only occurs if there’s nothing to highlight. So whatever we show here is only to make clear that there is a lane, so that the user doesn’t miscount when figuring out which lane to get into. A question mark would probably be more distracting than helpful for this purpose.

The most robust fallback would indeed be the largest subset of lane indications that are supported in a single icon. However, it’s a question of whether all this effort is necessary for the 19 affected roads in https://github.com/mapbox/navigation-ui-resources/pull/26#issuecomment-924307448, many of which are probably mistagged anyways.

The next most robust fallback would be to show any of the lane’s indications arbitrarily. There’s a risk of showing a left turn arrow where a right turn arrow would be _slightly_ less confusing, but it’s almost moot, so if you find it straightforward enough to select an indication at random, go for it.

File: libnavui-maneuver/src/main/java/com/mapbox/navigation/ui/maneuver/LaneIconProcessor.kt (hunk 2 above)
Comment (1ec5) [reply libnavui-maneuver/src/main/java/com/mapbox/navigation/ui/maneuver/LaneIconProcessor.kt:172]: [The `indications` documentation](https://docs.mapbox.com/api/navigation/directions/#lane-object) doesn’t reserve the right to emit an arbitrary value in the future as [some other enumeration-like fields](https://docs.mapbox.com/api/navigation/directions/#maneuver-types) do. But let’s suppose the Directions API has a bug where it fails to filter out OpenStreetMap’s `merge_to_left` value and emits `merge to left` as one of the indications. Depending on this codebase’s overall philosophy about failing early versus failing gracefully, it would be reasonable to either assert or fall back to the same fallback being discussed in https://github.com/mapbox/mapbox-navigation-android/pull/4854#discussion_r713131549. I’m leaning towards the latter for iOS.

Comment (abhishek1508) [reply libnavui-maneuver/src/main/java/com/mapbox/navigation/ui/maneuver/LaneIconProcessor.kt:172]: > it would be reasonable to either assert or fall back to the same fallback being discussed in #4854 (comment). I’m leaning towards the latter for iOS.

Good to know that one of the possible values can be `none` as well. I just want to say that whatever we end up choosing as the fallback, let's keep the behavior same on both the platforms. Either we both show `straight` or a `subset of the indication`.

Comment (LukasPaczos) [reply libnavui-maneuver/src/main/java/com/mapbox/navigation/ui/maneuver/LaneIconProcessor.kt]: I understand that we're aligned we'd rather fall back to something than throw an exception :+1:

File: libnavui-maneuver/src/main/java/com/mapbox/navigation/ui/maneuver/ManeuverProcessor.kt (hunk 4 above)
Comment (LukasPaczos) [reply libnavui-maneuver/src/main/java/com/mapbox/navigation/ui/maneuver/ManeuverProcessor.kt]: > As far as active is concerned; this can never be null.

It can be null if the component type is not `Lane`. This is my only concern.

As far as I see there's a `BannerComponent#type` which can be `Lane` (as visible in your capture), and only for this type the `BannerComponent#active` field is available, which we're not checking before coercing `active` to a non-nullable type.

Basically instead of checking `BannerSection#type == null`, we should be checking `BannerSection#components#type == Lane`, shouldn't we?

If there was in the future introduce a `BannerSection` which passed
```
subBanner.type() == null && subBanner.text().isEmpty()
``` 
but actually wasn't a `Lane`, we would start crashing.

Comment (abhishek1508) [reply libnavui-maneuver/src/main/java/com/mapbox/navigation/ui/maneuver/ManeuverProcessor.kt]: Added a check on `type == lane`.

/cc @LukasPaczos

File: libnavui-maneuver/src/main/java/com/mapbox/navigation/ui/maneuver/LaneIconProcessor.kt (hunk 1 above)
Comment (LukasPaczos) [reply libnavui-maneuver/src/main/java/com/mapbox/navigation/ui/maneuver/LaneIconProcessor.kt:180]: I understand that this is rare, and personally, I would still prefer selecting one of the actually viable turns but it's not a hill I'm willing to die on. If iOS is planning to use `straight` as a fallback, let's do that here as well.

File: libnavui-maneuver/src/test/java/com/mapbox/navigation/ui/maneuver/LaneIconsProcessorTest.kt [hunk 5]
```diff
@@ -0,0 +1,1776 @@
...
+import com.mapbox.api.directions.v5.models.ManeuverModifier.SLIGHT_LEFT
+import com.mapbox.api.directions.v5.models.ManeuverModifier.SLIGHT_RIGHT
+import com.mapbox.api.directions.v5.models.ManeuverModifier.STRAIGHT
+import com.mapbox.api.directions.v5.models.ManeuverModifier.UTURN
+import com.mapbox.navigation.ui.maneuver.model.LaneIconResources
+import com.mapbox.navigation.ui.maneuver.model.LaneIndicator
+import org.junit.Assert.assertEquals
+import org.junit.Test
+
+class LaneIconsProcessorTest {
```
Comment (LukasPaczos) [inline libnavui-maneuver/src/test/java/com/mapbox/navigation/ui/maneuver/LaneIconsProcessorTest.kt:16]: There are no tests for the left driving side, could you add some, especially for the critical edge cases?

**Review by LukasPaczos — APPROVED** [review]
Great work here @abhishek1508, and huge thanks for help @1ec5!

This generally LGTM with a request for more tests below.

Example of an improvement:
_before_
device-2021-09-22-193301

_after_
device-2021-09-22-193554
